jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/PathMappingsHandler.java

```diff
@@ -111,11 +113,49 @@ public boolean handle(Request request, Response response, Callback callback) thr
             return false;
         }
         Handler handler = matchedResource.getResource();
+        PathSpec pathSpec = matchedResource.getPathSpec();
         if (LOG.isDebugEnabled())
             LOG.debug("Matched {} to {} -> {}", pathInContext, matchedResource.getPathSpec(), handler);
-        boolean handled = handler.handle(request, response, callback);
+
+        PathSpecRequest pathSpecRequest = new PathSpecRequest(request, pathSpec);
+        boolean handled = handler.handle(pathSpecRequest, response, callback);
         if (LOG.isDebugEnabled())
             LOG.debug("Handled {} {} by {}", handled, pathInContext, handler);
         return handled;
     }
+
+    private static class PathSpecRequest extends Request.Wrapper
+    {
+        private final PathSpec pathSpec;
+        private final Context context;
+        private final MatchedPath matchedPath;
+
+        public PathSpecRequest(Request request, PathSpec pathSpec)
+        {
+            super(request);
+            this.pathSpec = pathSpec;
+            matchedPath = pathSpec.matched(request.getHttpURI().getCanonicalPath());
+            setAttribute(PathSpec.class.getName(), this.pathSpec);
+            this.context = new Context.Wrapper(request.getContext())
+            {
+                @Override
+                public String getContextPath()
+                {
+                    return matchedPath.getPathMatch();
+                }
+
+                @Override
+                public String getPathInContext(String canonicallyEncodedPath)
+                {
+                    return matchedPath.getPathInfo();
```

Comment:
Hmmm I'm not sure we can just return that pathInfo.  I think we have to do:
```suggestion
                    return Context.getPathInContext(getWrapped().getContext().getContextPath(), canonicallyEncodedPath);
```